# Portugese Bank Marketing Campaign term deposit classification

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [2]:
pd.set_option("display.max_columns",None)

data = pd.read_csv("Datasets/bank.csv",header=0,sep=None) 
data.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  duration  campaign  pdays  previous     poutcome  \
0   may         mon       261         1    999         0  nonexistent   
1   may         mon       149         1    999         0  nonexistent   
2   may         mon       226         1    999         0  nonexistent   
3   may         mon       151         1    999         0  nonexistent   
4   may         mon       307         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0           1.1          93.994          -36.4      4.857       5191.0  no  
1           1.1          93.994          -36.4      4.857       5191.0  no  
2           1.1          93.994          -36.4      4.857       5191.0  no  
3           1.1          93.994          -36.4      4.857       5191.0  no  
4           1.1          93.994          -36.4      4.857       5191.0  no

In [3]:
data.shape

(41188, 21)

In [4]:
data.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

In [5]:
data.rename(columns = {'y':'term_deposit'}, inplace = True)

In [6]:
data.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'term_deposit'],
      dtype='object')

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

- there are no null values in data

In [8]:
data.describe(include='all')

age     job  marital          education default housing  \
count   41188.00000   41188    41188              41188   41188   41188   
unique          NaN      12        4                  8       3       3   
top             NaN  admin.  married  university.degree      no     yes   
freq            NaN   10422    24928              12168   32588   21576   
mean       40.02406     NaN      NaN                NaN     NaN     NaN   
std        10.42125     NaN      NaN                NaN     NaN     NaN   
min        17.00000     NaN      NaN                NaN     NaN     NaN   
25%        32.00000     NaN      NaN                NaN     NaN     NaN   
50%        38.00000     NaN      NaN                NaN     NaN     NaN   
75%        47.00000     NaN      NaN                NaN     NaN     NaN   
max        98.00000     NaN      NaN                NaN     NaN     NaN   

         loan   contact  month day_of_week      duration      campaign  \
count   41188     41188  41188       41188  41188.000000  41188.000000   
unique      3         2     10           5           NaN           NaN   
top        no  cellular    may         thu           NaN           NaN   
freq    33950     26144  13769        8623           NaN           NaN   
mean      NaN       NaN    NaN         NaN    258.285010      2.567593   
std       NaN       NaN    NaN         NaN    259.279249      2.770014   
min       NaN       NaN    NaN         NaN      0.000000      1.000000   
25%       NaN       NaN    NaN         NaN    102.000000      1.000000   
50%       NaN       NaN    NaN         NaN    180.000000      2.000000   
75%       NaN       NaN    NaN         NaN    319.000000      3.000000   
max       NaN       NaN    NaN         NaN   4918.000000     56.000000   

               pdays      previous     poutcome  emp.var.rate  cons.price.idx  \
count   41188.000000  41188.000000        41188  41188.000000    41188.000000   
unique           NaN           NaN            3           NaN             NaN   
top              NaN           NaN  nonexistent           NaN             NaN   
freq             NaN           NaN        35563           NaN             NaN   
mean      962.475454      0.172963          NaN      0.081886       93.575664   
std       186.910907      0.494901          NaN      1.570960        0.578840   
min         0.000000      0.000000          NaN     -3.400000       92.201000   
25%       999.000000      0.000000          NaN     -1.800000       93.075000   
50%       999.000000      0.000000          NaN      1.100000       93.749000   
75%       999.000000      0.000000          NaN      1.400000       93.994000   
max       999.000000      7.000000          NaN      1.400000       94.767000   

        cons.conf.idx     euribor3m   nr.employed term_deposit  
count    41188.000000  41188.000000  41188.000000        41188  
unique            NaN           NaN           NaN            2  
top               NaN           NaN           NaN           no  
freq              NaN           NaN           NaN        36548  
mean       -40.502600      3.621291   5167.035911          NaN  
std          4.628198      1.734447     72.251528          NaN  
min        -50.800000      0.634000   4963.600000          NaN  
25%        -42.700000      1.344000   5099.100000          NaN  
50%        -41.800000      4.857000   5191.000000          NaN  
75%        -36.400000      4.961000   5228.100000          NaN  
max        -26.900000      5.045000   5228.100000          NaN

In [9]:
obj_col=[]
for x in data.columns[0:-1]:
    if data[x].dtype =='object':
        obj_col.append(x)

In [10]:
obj_col

['job',
 'marital',
 'education',
 'default',
 'housing',
 'loan',
 'contact',
 'month',
 'day_of_week',
 'poutcome']

## EDA - What does the primary analysis of several categorical features reveal?

In [11]:
data.groupby(['term_deposit','default'])['term_deposit'].count()

term_deposit  default
no            no         28391
              unknown     8154
              yes            3
yes           no          4197
              unknown      443
Name: term_deposit, dtype: int64

- as per data defaulters are not taking term_deposit, non defaulters are taking term deposit

In [12]:
data.groupby(['term_deposit','housing'])['term_deposit'].count()

term_deposit  housing
no            no         16596
              unknown      883
              yes        19069
yes           no          2026
              unknown      107
              yes         2507
Name: term_deposit, dtype: int64

- there are many customers who housing loan and dont take term deposit, there are less customers with
loan that take term deposit

In [13]:
data.groupby(['term_deposit','loan'])['term_deposit'].count()

term_deposit  loan   
no            no         30100
              unknown      883
              yes         5565
yes           no          3850
              unknown      107
              yes          683
Name: term_deposit, dtype: int64

- people who dont have personal loan still take up term deposit, very less cutomers who have personal loan take term deposits 683 customers

In [14]:
data.groupby(['term_deposit','job'])['term_deposit'].count()

term_deposit  job          
no            admin.           9070
              blue-collar      8616
              entrepreneur     1332
              housemaid         954
              management       2596
              retired          1286
              self-employed    1272
              services         3646
              student           600
              technician       6013
              unemployed        870
              unknown           293
yes           admin.           1352
              blue-collar       638
              entrepreneur      124
              housemaid         106
              management        328
              retired           434
              self-employed     149
              services          323
              student           275
              technician        730
              unemployed        144
              unknown            37
Name: term_deposit, dtype: int64

- More Admin job customers take term deposits as per data, 
least are taken by housemaids and entrepreneurs

In [15]:
data.groupby(['term_deposit','education'])['term_deposit'].count()

term_deposit  education          
no            basic.4y                3748
              basic.6y                2104
              basic.9y                5572
              high.school             8484
              illiterate                14
              professional.course     4648
              university.degree      10498
              unknown                 1480
yes           basic.4y                 428
              basic.6y                 188
              basic.9y                 473
              high.school             1031
              illiterate                 4
              professional.course      595
              university.degree       1670
              unknown                  251
Name: term_deposit, dtype: int64

- university.degree customers take more of term deposits followed by High school as per data, 
least are taken by illiterate

In [16]:
data.groupby(['term_deposit','marital'])['term_deposit'].count()

term_deposit  marital 
no            divorced     4136
              married     22396
              single       9948
              unknown        68
yes           divorced      476
              married      2532
              single       1620
              unknown        12
Name: term_deposit, dtype: int64

- married customers take more of term deposits followed by Single customers as per data, 
least are taken by divorced customers

In [17]:
data.groupby(['term_deposit','poutcome'])['term_deposit'].count()

term_deposit  poutcome   
no            failure         3647
              nonexistent    32422
              success          479
yes           failure          605
              nonexistent     3141
              success          894
Name: term_deposit, dtype: int64

- nonexistent customers take more of term deposits followed by Success customers as per data of previous marketing campaign outcome, 
there are few customers who had success outcome last campaign but have not taken term deposit this time

In [18]:
data.groupby(['term_deposit','day_of_week'])['term_deposit'].count()

term_deposit  day_of_week
no            fri            6981
              mon            7667
              thu            7578
              tue            7137
              wed            7185
yes           fri             846
              mon             847
              thu            1045
              tue             953
              wed             949
Name: term_deposit, dtype: int64

- last contact day of week thursday customers have taken maximum term deposits

In [19]:
data.groupby(['term_deposit','month'])['term_deposit'].count()

term_deposit  month
no            apr       2093
              aug       5523
              dec         93
              jul       6525
              jun       4759
              mar        270
              may      12883
              nov       3685
              oct        403
              sep        314
yes           apr        539
              aug        655
              dec         89
              jul        649
              jun        559
              mar        276
              may        886
              nov        416
              oct        315
              sep        256
Name: term_deposit, dtype: int64

- last contact month at May customers have taken maximum term deposits, least are taken contact at December customers

In [20]:
data.groupby(['term_deposit','contact'])['term_deposit'].count()

term_deposit  contact  
no            cellular     22291
              telephone    14257
yes           cellular      3853
              telephone      787
Name: term_deposit, dtype: int64

- cellular contact mode customers have taken maximum term deposits than telephone contact

## Checking analomalies in Data

In [21]:
data[data.campaign==56.000000]

age     job  marital          education  default  housing     loan  \
4107   32  admin.  married  university.degree  unknown  unknown  unknown   

        contact month day_of_week  duration  campaign  pdays  previous  \
4107  telephone   may         mon       261        56    999         0   

         poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
4107  nonexistent           1.1          93.994          -36.4      4.858   

      nr.employed term_deposit  
4107       5191.0           no

In [22]:
data.campaign.unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 19, 18, 23, 14,
       22, 25, 16, 17, 15, 20, 56, 39, 35, 42, 28, 26, 27, 32, 21, 24, 29,
       31, 30, 41, 37, 40, 33, 34, 43], dtype=int64)

In [23]:
data.pdays.unique()

array([999,   6,   4,   3,   5,   1,   0,  10,   7,   8,   9,  11,   2,
        12,  13,  14,  15,  16,  21,  17,  18,  22,  25,  26,  19,  27,
        20], dtype=int64)

In [24]:
data.pdays.value_counts()

pdays
999    39673
3        439
6        412
4        118
9         64
2         61
7         60
12        58
10        52
5         46
13        36
11        28
1         26
15        24
14        20
8         18
0         15
16        11
17         8
18         7
22         3
19         3
21         2
25         1
26         1
27         1
20         1
Name: count, dtype: int64

In [25]:
data.previous.unique()

array([0, 1, 2, 3, 4, 5, 6, 7], dtype=int64)

In [26]:
data.duration.unique()

array([ 261,  149,  226, ..., 1246, 1556, 1868], dtype=int64)

In [27]:
data['cons.price.idx'].unique()

array([93.994, 94.465, 93.918, 93.444, 93.798, 93.2  , 92.756, 92.843,
       93.075, 92.893, 92.963, 92.469, 92.201, 92.379, 92.431, 92.649,
       92.713, 93.369, 93.749, 93.876, 94.055, 94.215, 94.027, 94.199,
       94.601, 94.767])

In [28]:
data['cons.conf.idx'].unique()

array([-36.4, -41.8, -42.7, -36.1, -40.4, -42. , -45.9, -50. , -47.1,
       -46.2, -40.8, -33.6, -31.4, -29.8, -26.9, -30.1, -33. , -34.8,
       -34.6, -40. , -39.8, -40.3, -38.3, -37.5, -49.5, -50.8])

In [29]:
data[data.pdays==0]

age          job   marital            education  default housing loan  \
25497   31   unemployed    single  professional.course       no     yes   no   
25996   35   management   married    university.degree       no     yes  yes   
30238   51   technician   married  professional.course       no     yes   no   
30314   36       admin.   married    university.degree       no     yes   no   
30417   30       admin.    single    university.degree       no     yes   no   
30688   35  blue-collar   married             basic.6y  unknown     yes   no   
40360   48       admin.   married    university.degree       no     yes  yes   
40364   27      student    single    university.degree       no     yes   no   
40365   28      student    single             basic.9y       no     yes   no   
40376   27      student    single    university.degree       no     yes   no   
40380   48       admin.    single    university.degree       no     yes   no   
40382   30       admin.    single    university.degree       no     yes   no   
40384   44     services  divorced             basic.6y       no     yes   no   
40387   41       admin.  divorced          high.school       no     yes  yes   
40534   30   technician   married  professional.course       no      no   no   

         contact month day_of_week  duration  campaign  pdays  previous  \
25497   cellular   nov         wed       192         1      0         1   
25996  telephone   nov         wed       377         3      0         2   
30238   cellular   apr         thu       687         1      0         1   
30314   cellular   apr         thu       483         1      0         1   
30417   cellular   apr         thu       502         1      0         1   
30688   cellular   may         tue       214         1      0         1   
40360   cellular   aug         wed       288         1      0         3   
40364   cellular   aug         wed       651         1      0         3   
40365   cellular   aug         wed       178         1      0         1   
40376  telephone   aug         wed       153         1      0         5   
40380   cellular   aug         wed       118         1      0         2   
40382   cellular   aug         wed       191         1      0         2   
40384   cellular   aug         wed       153         3      0         1   
40387   cellular   aug         thu       231         2      0         2   
40534  telephone   aug         wed         5         1      0         4   

      poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
25497  success          -0.1          93.200          -42.0      4.120   
25996  success          -0.1          93.200          -42.0      4.120   
30238  success          -1.8          93.075          -47.1      1.365   
30314  success          -1.8          93.075          -47.1      1.365   
30417  success          -1.8          93.075          -47.1      1.365   
30688  success          -1.8          92.893          -46.2      1.344   
40360  success          -1.7          94.027          -38.3      0.900   
40364  success          -1.7          94.027          -38.3      0.900   
40365  success          -1.7          94.027          -38.3      0.900   
40376  success          -1.7          94.027          -38.3      0.900   
40380  success          -1.7          94.027          -38.3      0.900   
40382  success          -1.7          94.027          -38.3      0.900   
40384  success          -1.7          94.027          -38.3      0.900   
40387  success          -1.7          94.027          -38.3      0.904   
40534  success          -1.7          94.027          -38.3      0.890   

       nr.employed term_deposit  
25497       5195.8           no  
25996       5195.8           no  
30238       5099.1          yes  
30314       5099.1          yes  
30417       5099.1          yes  
30688       5099.1           no  
40360       4991.6          yes  
40364       4991.6          yes  
40365       4991.6          yes  


### Conclusion
- i dont see any anomalies in data

## Unique values in every column

In [30]:
for i in data.columns:
    print({i:data[i].unique()})

{'age': array([56, 57, 37, 40, 45, 59, 41, 24, 25, 29, 35, 54, 46, 50, 39, 30, 55,
       49, 34, 52, 58, 32, 38, 44, 42, 60, 53, 47, 51, 48, 33, 31, 43, 36,
       28, 27, 26, 22, 23, 20, 21, 61, 19, 18, 70, 66, 76, 67, 73, 88, 95,
       77, 68, 75, 63, 80, 62, 65, 72, 82, 64, 71, 69, 78, 85, 79, 83, 81,
       74, 17, 87, 91, 86, 98, 94, 84, 92, 89], dtype=int64)}
{'job': array(['housemaid', 'services', 'admin.', 'blue-collar', 'technician',
       'retired', 'management', 'unemployed', 'self-employed', 'unknown',
       'entrepreneur', 'student'], dtype=object)}
{'marital': array(['married', 'single', 'divorced', 'unknown'], dtype=object)}
{'education': array(['basic.4y', 'high.school', 'basic.6y', 'basic.9y',
       'professional.course', 'unknown', 'university.degree',
       'illiterate'], dtype=object)}
{'default': array(['no', 'unknown', 'yes'], dtype=object)}
{'housing': array(['no', 'yes', 'unknown'], dtype=object)}
{'loan': array(['no', 'yes', 'unknown'], dtype=object)}
{'c

In [31]:
data.term_deposit.replace({'no':0,'yes':1},inplace=True)

In [32]:
data.term_deposit.unique()

array([0, 1], dtype=int64)

## preprocessing the data

In [33]:
data.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
term_deposit      0
dtype: int64

- there are no missing values in the data

### Label Encoding the categorical variables

In [34]:
colname=[]
for x in data.columns:
    if data[x].dtype=='object':
        colname.append(x)
colname

['job',
 'marital',
 'education',
 'default',
 'housing',
 'loan',
 'contact',
 'month',
 'day_of_week',
 'poutcome']

In [35]:
col1= ['job',
 'marital',
 'default',
 'housing',
 'loan',
 'contact',
 'poutcome']

In [36]:
# For preprocessing the data
from sklearn.preprocessing import LabelEncoder
 
le=LabelEncoder()
 
for x in col1:
    data[x]=le.fit_transform(data[x]) #mandatory
    

    le_name_mapping = dict(zip(le.classes_, le.transform(le.classes_))) #optional check labels
    print('Feature', x)
    print('mapping', le_name_mapping)


Feature job
mapping {'admin.': 0, 'blue-collar': 1, 'entrepreneur': 2, 'housemaid': 3, 'management': 4, 'retired': 5, 'self-employed': 6, 'services': 7, 'student': 8, 'technician': 9, 'unemployed': 10, 'unknown': 11}
Feature marital
mapping {'divorced': 0, 'married': 1, 'single': 2, 'unknown': 3}
Feature default
mapping {'no': 0, 'unknown': 1, 'yes': 2}
Feature housing
mapping {'no': 0, 'unknown': 1, 'yes': 2}
Feature loan
mapping {'no': 0, 'unknown': 1, 'yes': 2}
Feature contact
mapping {'cellular': 0, 'telephone': 1}
Feature poutcome
mapping {'failure': 0, 'nonexistent': 1, 'success': 2}


In [37]:
data.education.unique()

array(['basic.4y', 'high.school', 'basic.6y', 'basic.9y',
       'professional.course', 'unknown', 'university.degree',
       'illiterate'], dtype=object)

In [38]:
data.education.replace({'unknown':0,'illiterate':1,'basic.4y':2,'basic.6y':3,'basic.9y':4,'high.school':5,'university.degree':6,'professional.course':7},inplace=True)

In [39]:
data.month.unique()

array(['may', 'jun', 'jul', 'aug', 'oct', 'nov', 'dec', 'mar', 'apr',
       'sep'], dtype=object)

In [40]:
data.month.replace({'mar':3,'apr':4,'may':5,'jun':6,'jul':7,'aug':8,'sep':9,'oct':10,'nov':11,'dec':12},inplace=True)

In [41]:
data.day_of_week.unique()

array(['mon', 'tue', 'wed', 'thu', 'fri'], dtype=object)

In [42]:
data.day_of_week.replace({'mon':1,'tue':2,'wed':3,'thu':4,'fri':5},inplace=True)

In [43]:
data.head()

age  job  marital  education  default  housing  loan  contact  month  \
0   56    3        1          2        0        0     0        1      5   
1   57    7        1          5        1        0     0        1      5   
2   37    7        1          5        0        2     0        1      5   
3   40    0        1          3        0        0     0        1      5   
4   56    7        1          5        0        0     2        1      5   

   day_of_week  duration  campaign  pdays  previous  poutcome  emp.var.rate  \
0            1       261         1    999         0         1           1.1   
1            1       149         1    999         0         1           1.1   
2            1       226         1    999         0         1           1.1   
3            1       151         1    999         0         1           1.1   
4            1       307         1    999         0         1           1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed  term_deposit  
0          93.994          -36.4      4.857       5191.0             0  
1          93.994          -36.4      4.857       5191.0             0  
2          93.994          -36.4      4.857       5191.0             0  
3          93.994          -36.4      4.857       5191.0             0  
4          93.994          -36.4      4.857       5191.0             0

## Random Forest Feature selection

In [44]:
X= data.iloc[:,:-1]
Y=data.term_deposit

In [45]:
X.shape

(41188, 20)

In [46]:
Y.shape

(41188,)

- duration attribute highly affects the output target (e.g., if duration=0 then y="no"). Yet, the 
duration is not known before a call is performed. Also, after the end of the call y is obviously 
known. Thus,this input should only be included for benchmark purposes and should be discarded if the 
intention is to have a realistic predictive model.

In [47]:
X.drop('duration',axis=1,inplace=True)

In [48]:
X.shape

(41188, 19)

In [49]:
from sklearn.ensemble import RandomForestClassifier
#create a model object
model_RandomForest = RandomForestClassifier(random_state=10,n_estimators=100,bootstrap=True,n_jobs=-1)
# train the model object
model_RandomForest.fit(X,Y) #trainig phase is building the DTree Trees 100

RandomForestClassifier(n_jobs=-1, random_state=10)

In [50]:
sample=pd.DataFrame()
 
sample["Feature"]=X.columns
sample["Feature Importance"]=model_RandomForest.feature_importances_
 
sample.sort_values("Feature Importance",ascending=False)

Feature  Feature Importance
0              age            0.186383
17       euribor3m            0.142383
10        campaign            0.087808
1              job            0.085556
3        education            0.075986
9      day_of_week            0.058910
18     nr.employed            0.057340
11           pdays            0.043618
2          marital            0.041074
5          housing            0.037952
16   cons.conf.idx            0.028221
13        poutcome            0.026240
6             loan            0.025972
15  cons.price.idx            0.021735
14    emp.var.rate            0.019733
12        previous            0.016793
8            month            0.015962
7          contact            0.014563
4          default            0.013769

- age and euribor3m attributes have higher importance in predicting term deposit
- previous,month,contact,default attributes have lower importance in predicting term deposit

- for now i have dropped duration and kept other feature further steps i will tune the models by feature selection

## Standardization using Standard Scaler

In [51]:
from sklearn.preprocessing import StandardScaler

scaler= StandardScaler()

scaler.fit(X)

X= scaler.transform(X)

In [52]:
X

array([[ 1.53303429, -0.20157925, -0.2837415 , ...,  0.88644656,
         0.71245988,  0.33167991],
       [ 1.62899323,  0.91122681, -0.2837415 , ...,  0.88644656,
         0.71245988,  0.33167991],
       [-0.29018564,  0.91122681, -0.2837415 , ...,  0.88644656,
         0.71245988,  0.33167991],
       ...,
       [ 1.53303429,  0.35482378, -0.2837415 , ..., -2.22495344,
        -1.49518647, -2.8156966 ],
       [ 0.38152696,  1.46762984, -0.2837415 , ..., -2.22495344,
        -1.49518647, -2.8156966 ],
       [ 3.26029527,  0.35482378, -0.2837415 , ..., -2.22495344,
        -1.49518647, -2.8156966 ]])

In [53]:
from sklearn.model_selection import train_test_split
  
# split into 80:20 ratio
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, 
                                                    random_state = 10)

In [54]:
print(X_train.shape)
print(Y_train.shape)
print(X_test.shape)
print(Y_test.shape)

(28831, 19)
(28831,)
(12357, 19)
(12357,)


## SMOTE

In [55]:
Y_train.value_counts() #unbalanced class Y perform SMOTE

term_deposit
0    25613
1     3218
Name: count, dtype: int64

In [56]:
print("Before OverSampling, counts of label '1': ", (sum(Y_train == 1)))
print("Before OverSampling, counts of label '0': ", (sum(Y_train == 0)))
  
# import SMOTE from imblearn library
from imblearn.over_sampling import SMOTE
sm = SMOTE(random_state = 10,k_neighbors=5)
X_train_res, Y_train_res = sm.fit_resample(X_train, Y_train)
  
print('After OverSampling, the shape of train_X: ', (X_train_res.shape))
print('After OverSampling, the shape of train_y: ', (Y_train_res.shape))
  
print("After OverSampling, counts of label '1': ", (sum(Y_train_res == 1)))
print("After OverSampling, counts of label '0': ", (sum(Y_train_res == 0)))

Before OverSampling, counts of label '1':  3218
Before OverSampling, counts of label '0':  25613
After OverSampling, the shape of train_X:  (51226, 19)
After OverSampling, the shape of train_y:  (51226,)
After OverSampling, counts of label '1':  25613
After OverSampling, counts of label '0':  25613


## Supervised Learning models

In [57]:
from sklearn.linear_model import LogisticRegression
#create a model object
classifier = LogisticRegression()
# train the model object
classifier.fit(X_train,Y_train)

#print intercept and coefficients
print(classifier.intercept_)
print(classifier.coef_)

[-2.49370894]
[[ 6.04429014e-02  1.08627749e-02  6.49664998e-02  2.84875381e-02
  -1.39444245e-01 -9.90747765e-03  2.51135617e-04 -4.78900123e-01
  -1.05501672e-01  5.41569422e-02 -1.44885448e-01 -1.90250344e-01
  -4.11062007e-02  1.60822216e-01 -1.25153097e+00  6.13472037e-01
   2.03031064e-01  5.32876371e-01 -3.84340501e-01]]


In [58]:
#pair x column names and its slope coeffs
print(list(zip(data.drop(['duration','term_deposit'],axis=1).columns,classifier.coef_.ravel())))

[('age', 0.06044290142354115), ('job', 0.010862774909403931), ('marital', 0.06496649981381569), ('education', 0.028487538110035064), ('default', -0.13944424522717422), ('housing', -0.009907477649398049), ('loan', 0.000251135617418765), ('contact', -0.4789001232846601), ('month', -0.10550167221786184), ('day_of_week', 0.054156942248926095), ('campaign', -0.14488544790543054), ('pdays', -0.1902503435684248), ('previous', -0.04110620068946247), ('poutcome', 0.16082221564085633), ('emp.var.rate', -1.251530974749395), ('cons.price.idx', 0.6134720373254986), ('cons.conf.idx', 0.2030310641261809), ('euribor3m', 0.5328763711765848), ('nr.employed', -0.38434050112511214)]


In [59]:
Y_pred_logistic=classifier.predict(X_test)
print(Y_pred_logistic) #term deposit

[0 0 1 ... 0 0 0]


In [60]:
Y_pred_logistic.shape

(12357,)

In [61]:
#prob matrix which displays prob of every obs to belong to class 0 or 1
y_pred_prob_matrix=classifier.predict_proba(X_test)
print(y_pred_prob_matrix)

[[0.92882886 0.07117114]
 [0.93777476 0.06222524]
 [0.26333527 0.73666473]
 ...
 [0.97484776 0.02515224]
 [0.92635031 0.07364969]
 [0.9512903  0.0487097 ]]


In [62]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred_logistic)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred_logistic))
 
acc=accuracy_score(Y_test,Y_pred_logistic)
print("Accuracy of the model: ",acc)

[[10787   148]
 [ 1129   293]]
Classification report: 
              precision    recall  f1-score   support

           0       0.91      0.99      0.94     10935
           1       0.66      0.21      0.31      1422

    accuracy                           0.90     12357
   macro avg       0.78      0.60      0.63     12357
weighted avg       0.88      0.90      0.87     12357

Accuracy of the model:  0.8966577648296512


- the recall for class 1 is very low 0.21, from 1422 observations of class 1 only 293 are correctly predicted
- the recall for class 0 is good
- the total correct predictions is 11080 out of 12357 observations,thus the accuracy is high.
- we need to improve class 1 recall which is our class of importance that is bank customers subscribe term deposit
- thus we will tune the base model

In [63]:
from sklearn.tree import DecisionTreeClassifier
#create a model object
model_DecisionTree = DecisionTreeClassifier(random_state=10,criterion='gini')
# train the model object
model_DecisionTree.fit(X_train,Y_train) #trainig phase is building the DTree fit Tree

Y_pred_DT=model_DecisionTree.predict(X_test)
print(Y_pred_DT)

[0 0 1 ... 0 0 0]


In [64]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred_DT)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred_DT))
 
acc=accuracy_score(Y_test, Y_pred_DT)
print("Accuracy of the model: ",acc)

[[9824 1111]
 [ 936  486]]
Classification report: 
              precision    recall  f1-score   support

           0       0.91      0.90      0.91     10935
           1       0.30      0.34      0.32      1422

    accuracy                           0.83     12357
   macro avg       0.61      0.62      0.61     12357
weighted avg       0.84      0.83      0.84     12357

Accuracy of the model:  0.8343449057214535


- the recall for class 1 has increased in DT compared to Logistic Regression which is good but accuracy has dropped
- the recall for class 1 is very low 0.34, from 1422 observations of class 1 only 486 are correctly predicted
- the recall for class 0 is good
- the total correct predictions is 10310 out of 12357 observations,thus the accuracy is good.
- we need to improve class 1 recall which is our class of importance that is bank customers subscribe term deposit
- thus we will tune the base model

In [65]:
from sklearn.ensemble import RandomForestClassifier
#create a model object
model_RandomForest = RandomForestClassifier(random_state=10,n_estimators=100,bootstrap=True,n_jobs=-1)
# train the model object
model_RandomForest.fit(X_train,Y_train) #training phase is building the DTree Trees 100

RandomForestClassifier(n_jobs=-1, random_state=10)

In [66]:
y_pred_forest=model_RandomForest.predict(X_test)
print(y_pred_forest)

[0 0 1 ... 0 0 0]


In [67]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,y_pred_forest)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,y_pred_forest))
 
acc=accuracy_score(Y_test, y_pred_forest)
print("Accuracy of the model: ",acc)

[[10610   325]
 [ 1036   386]]
Classification report: 
              precision    recall  f1-score   support

           0       0.91      0.97      0.94     10935
           1       0.54      0.27      0.36      1422

    accuracy                           0.89     12357
   macro avg       0.73      0.62      0.65     12357
weighted avg       0.87      0.89      0.87     12357

Accuracy of the model:  0.8898599983814842


- the recall for class 1 is very low 0.27, from 1422 observations of class 1 only 386 are correctly predicted
- the recall for class 0 is good
- the total correct predictions is 10996 out of 12357 observations,thus the accuracy is good.
- we need to improve class 1 recall which is our class of importance that is bank customers subscribe term deposit
- thus we will tune the base model

## Tabulate the performance metrics of all the above models (Base models)

In [68]:
evaluation_base_model=pd.DataFrame()
evaluation_base_model['Model']=['logistic regression','decision tree','Random Forest']
evaluation_base_model['Accuracy_score']=[0.8966577648296512,0.8343449057214535,0.8898599983814842]
evaluation_base_model['Recall_class0']=[0.99,0.90,0.97]
evaluation_base_model['Recall_class1']=[0.21,0.34,0.27]
evaluation_base_model['f1_score_class0']=[0.94,0.91,0.94]
evaluation_base_model['f1_score_class1']=[0.31,0.32,0.36]


In [69]:
evaluation_base_model

Model  Accuracy_score  Recall_class0  Recall_class1  \
0  logistic regression        0.896658           0.99           0.21   
1        decision tree        0.834345           0.90           0.34   
2        Random Forest        0.889860           0.97           0.27   

   f1_score_class0  f1_score_class1  
0             0.94             0.31  
1             0.91             0.32  
2             0.94             0.36

## Tunning the base models

### Tunning logistic regression base model

In [70]:
#Lets find the best treshold for sigmoid function for model prob matrix 
for a in np.arange(0.4,0.51,0.01):
    predict_mine = np.where(y_pred_prob_matrix[:,1] > a, 1, 0)
    cfm=confusion_matrix(Y_test, predict_mine)
    total_err=cfm[0,1]+cfm[1,0]
    print("Errors at threshold ", a, ":",total_err, " , type 2 error :", 
          cfm[1,0]," , type 1 error:", cfm[0,1])

Errors at threshold  0.4 : 1313  , type 2 error : 996  , type 1 error: 317
Errors at threshold  0.41000000000000003 : 1312  , type 2 error : 1004  , type 1 error: 308
Errors at threshold  0.42000000000000004 : 1312  , type 2 error : 1018  , type 1 error: 294
Errors at threshold  0.43000000000000005 : 1311  , type 2 error : 1033  , type 1 error: 278
Errors at threshold  0.44000000000000006 : 1303  , type 2 error : 1040  , type 1 error: 263
Errors at threshold  0.45000000000000007 : 1299  , type 2 error : 1056  , type 1 error: 243
Errors at threshold  0.4600000000000001 : 1296  , type 2 error : 1065  , type 1 error: 231
Errors at threshold  0.4700000000000001 : 1278  , type 2 error : 1073  , type 1 error: 205
Errors at threshold  0.4800000000000001 : 1280  , type 2 error : 1094  , type 1 error: 186
Errors at threshold  0.4900000000000001 : 1283  , type 2 error : 1114  , type 1 error: 169
Errors at threshold  0.5000000000000001 : 1277  , type 2 error : 1129  , type 1 error: 148


- 0.4 is the best threshold for model prob matrix as type 2 error : 996 is the least

In [71]:
y_pred_class=[]
for value in y_pred_prob_matrix[:,1]:
    if value > 0.4:
        y_pred_class.append(1)
    else:
        y_pred_class.append(0)
#print(y_pred_class)

In [72]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,y_pred_class)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,y_pred_class))
 
acc=accuracy_score(Y_test, y_pred_class)
print("Accuracy of the model: ",acc)

[[10618   317]
 [  996   426]]
Classification report: 
              precision    recall  f1-score   support

           0       0.91      0.97      0.94     10935
           1       0.57      0.30      0.39      1422

    accuracy                           0.89     12357
   macro avg       0.74      0.64      0.67     12357
weighted avg       0.88      0.89      0.88     12357

Accuracy of the model:  0.8937444363518653


- the recall for class 1 has improved after tunning 0.30 so model has improved

### Tunning Decision Tree base model

In [73]:
#Prunning the tree

from sklearn.tree import DecisionTreeClassifier
#create a model object
model_DecisionTree1 = DecisionTreeClassifier(random_state=10,criterion='gini',splitter='best',max_depth=25,max_features=0.9,min_samples_split=5) #max_features=0.6
# train the model object
model_DecisionTree1.fit(X_train,Y_train) #trainig phase is building the DTree fit Tree

y_pred_DT1=model_DecisionTree1.predict(X_test)
print(y_pred_DT1)

[0 0 1 ... 0 0 0]


In [74]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,y_pred_DT1)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,y_pred_DT1))
 
acc=accuracy_score(Y_test, y_pred_DT1)
print("Accuracy of the model: ",acc)

[[10167   768]
 [  984   438]]
Classification report: 
              precision    recall  f1-score   support

           0       0.91      0.93      0.92     10935
           1       0.36      0.31      0.33      1422

    accuracy                           0.86     12357
   macro avg       0.64      0.62      0.63     12357
weighted avg       0.85      0.86      0.85     12357

Accuracy of the model:  0.8582180140810877


- even after tunning different combinations the Base Decision Tree model is performing better than the prunned Tree.
- Thus Base Decision Tree is a model, tunning not required

### Tunning Random forest base model

In [75]:
from sklearn.ensemble import RandomForestClassifier
#create a model object
model_RandomForest1 = RandomForestClassifier(random_state=10,n_estimators=227,bootstrap=True,n_jobs=-1,max_depth=30,max_features=0.90)
# train the model object
model_RandomForest1.fit(X_train,Y_train) #training phase is building the DTree Trees 100

y_pred_forest1=model_RandomForest1.predict(X_test)
#print(y_pred_forest1)

In [76]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,y_pred_forest1)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,y_pred_forest1))
 
acc=accuracy_score(Y_test, y_pred_forest1)
print("Accuracy of the model: ",acc)

[[10584   351]
 [ 1019   403]]
Classification report: 
              precision    recall  f1-score   support

           0       0.91      0.97      0.94     10935
           1       0.53      0.28      0.37      1422

    accuracy                           0.89     12357
   macro avg       0.72      0.63      0.65     12357
weighted avg       0.87      0.89      0.87     12357

Accuracy of the model:  0.8891316662620377


- base DTree model is good recall 0.34, 936 misclassified 

In [77]:
print(list(zip(data.drop(['duration','term_deposit'],axis=1),model_DecisionTree.feature_importances_)))

[('age', 0.16629869425385185), ('job', 0.07594813455933569), ('marital', 0.04292417321663726), ('education', 0.08160477162839966), ('default', 0.009173198471986938), ('housing', 0.03839895057647618), ('loan', 0.03348660035779555), ('contact', 0.011313137701701932), ('month', 0.019539473169250792), ('day_of_week', 0.05657518319938974), ('campaign', 0.09012770416845584), ('pdays', 0.013856563633733554), ('previous', 0.019463952232552005), ('poutcome', 0.0384833584981024), ('emp.var.rate', 0.004554447840748484), ('cons.price.idx', 0.012364095699618554), ('cons.conf.idx', 0.012696577487413192), ('euribor3m', 0.10542275599302638), ('nr.employed', 0.16776822731152408)]


## Tabulate the performance metrics of all the above models After tunning

In [78]:
evaluation_tunned_model=pd.DataFrame()
evaluation_tunned_model['Model_tunned']=['logistic regression','decision tree','Random Forest']
evaluation_tunned_model['Accuracy_score']=[0.8937444363518653,0.8582180140810877,0.8891316662620377]
evaluation_tunned_model['Recall_class0']=[0.97,0.93,0.97]
evaluation_tunned_model['Recall_class1']=[0.30,0.31,0.28]
evaluation_tunned_model['f1_score_class0']=[0.94,0.92,0.94]
evaluation_tunned_model['f1_score_class1']=[0.39,0.33,0.37]

In [79]:
evaluation_base_model

Model  Accuracy_score  Recall_class0  Recall_class1  \
0  logistic regression        0.896658           0.99           0.21   
1        decision tree        0.834345           0.90           0.34   
2        Random Forest        0.889860           0.97           0.27   

   f1_score_class0  f1_score_class1  
0             0.94             0.31  
1             0.91             0.32  
2             0.94             0.36

In [80]:
evaluation_tunned_model

Model_tunned  Accuracy_score  Recall_class0  Recall_class1  \
0  logistic regression        0.893744           0.97           0.30   
1        decision tree        0.858218           0.93           0.31   
2        Random Forest        0.889132           0.97           0.28   

   f1_score_class0  f1_score_class1  
0             0.94             0.39  
1             0.92             0.33  
2             0.94             0.37

# Conclusion
- The base decision tree model is a better model for predicting term deposit subscription of bank customers as recall for class 1 is 0.34
  and accuracy is 0.834345 i.e. 83.43% which is good model
- Other models recall for class 1 is poor.
- Thus base decision tree is final selected model